### PR TITLE
xcc: fix clang build issue

### PR DESCRIPTION
**Pull request by dcpleung**
This fixes issues when building using XCC-Clang toolchain.

### CODE REVIEW

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
btw, I am seeing this with gcc based xcc:

```
[28/114] Building C object zephyr/CMakeFiles/zephyr.dir/drivers/interrupt_controller/intc_cavs.c.obj
FAILED: zephyr/CMakeFiles/zephyr.dir/drivers/interrupt_controller/intc_cavs.c.obj
ccache /opt/xtensa/XtDevTools/install/tools/RF-2016.4-linux/XtensaTools/bin/xt-xcc -DBUILD_VERSION=v2.5.0-rc2-21-gb49292634fc9 -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -I../kernel/include -I../arch/xtensa/include -I../include -Izephyr/include/generated -I../soc/xtensa/intel_s1000 -I../soc/xtensa/intel_s1000/include -isystem ../lib/libc/minimal/include -Os -imacros /home/nashif/Work/zephyrproject/zephyr/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -imacros/home/nashif/Work/zephyrproject/zephyr/include/toolchain/xcc_missing_defs.h -fms-extensions -imacros /home/nashif/Work/zephyrproject/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Werror=implicit-int -fno-pic -fno-strict-overflow -ffunction-sections -fdata-sections -mlongcalls -std=c99 -MD -MT zephyr/CMakeFiles/zephyr.dir/drivers/interrupt_controller/intc_cavs.c.obj -MF zephyr/CMakeFiles/zephyr.dir/drivers/interrupt_controller/intc_cavs.c.obj.d -o zephyr/CMakeFiles/zephyr.dir/drivers/interrupt_controller/intc_cavs.c.obj -c /home/nashif/Work/zephyrproject/zephyr/drivers/interrupt_controller/intc_cavs.c
/home/nashif/Work/zephyrproject/zephyr/drivers/interrupt_controller/intc_cavs.c:165: error: nested redefinition of ‘enum __build_assert_enum165’
/home/nashif/Work/zephyrproject/zephyr/drivers/interrupt_controller/intc_cavs.c:165: error: redeclaration of ‘enum __build_assert_enum165’
/home/nashif/Work/zephyrproject/zephyr/drivers/interrupt_controller/intc_cavs.c:165: error: redeclaration of enumerator ‘__build_assert165’
/home/nashif/Work/zephyrproject/zephyr/drivers/interrupt_controller/intc_cavs.c:165: error: previous definition of ‘__build_assert165’ was here
/home/nashif/Work/zephyrproject/zephyr/drivers/interrupt_controller/intc_cavs.c:165: error: nested redefinition of ‘enum __build_assert_enum165’
/home/nashif/Work/zephyrproject/zephyr/drivers/interrupt_controller/intc_cavs.c:165: error: redeclaration of ‘enum __build_assert_enum165’
/home/nashif/Work/zephyrproject/zephyr/drivers/interrupt_controller/intc_cavs.c:165: error: redeclaration of enumerator ‘__build_assert165’
/home/nashif/Work/zephyrproject/zephyr/drivers/interrupt_controller/intc_cavs.c:165: error: previous definition of ‘__build_assert165’ was here
/home/nashif/Work/zephyrproject/zephyr/drivers/interrupt_controller/intc_cavs.c:165: error: nested redefinition of ‘enum __build_assert_enum165’
/home/nashif/Work/zephyrproject/zephyr/drivers/interrupt_controller/intc_cavs.c:165: error: redeclaration of ‘enum __build_assert_enum165’
/home/nashif/Work/zephyrproject/zephyr/drivers/interrupt_controller/intc_cavs.c:165: error: redeclaration of enumerator ‘__build_assert165’
/home/nashif/Work/zephyrproject/zephyr/drivers/interrupt_controller/intc_cavs.c:165: error: previous definition of ‘__build_assert165’ was here
[49/114] Building C object zephyr/arch/arch/xtensa/core/CMakeFiles/arch__xtensa__core.dir/xtensa-asm2.c.obj
ninja: build stopped: subcommand failed.
```

**Comment by dcpleung** [comment]
GCC-based XCC doesn't have `__COUNTER__` to be used for build assertion. We are substituting it with `__LINE__`. But with all the device tree stuff, looping through device tree nodes are all from one line, hence duplicate definitions.

File: arch/xtensa/core/xtensa-asm2-util.S [hunk 1]
```diff
@@ -147,35 +147,63 @@ _high_restore_done:
  * though...
  */
 
-#define RCTX_EPC_REG		_CONCAT(EPC, XCHAL_NUM_INTLEVELS)
-#define RCTX_EPS_REG		_CONCAT(EPS, XCHAL_NUM_INTLEVELS)
+/*
+ * XCC doesn't like _CONCAT() so we need to spell out
+ * every single one.
```
Comment (andyross) [inline arch/xtensa/core/xtensa-asm2-util.S:152]: Hm... is this something to do with the clang preprocessor?  Seems like we're going to hit this elsewhere if _CONCAT doesn't work on clang.

**Review by andyross — APPROVED** [review]

File: arch/xtensa/core/xtensa-asm2-util.S (hunk 1 above)
Comment (dcpleung) [reply arch/xtensa/core/xtensa-asm2-util.S:152]: This seems to be only limited to the assembler. We have C code using _CONCAT and XCC-Clang never complains.

**Pull request merged**